BaseLib/ConfigTreeNew.h

```diff
@@ -0,0 +1,392 @@
+            return _it == other._it;
+        }
+
+        bool operator!=(SubtreeIterator const& other) const {
+            return _it != other._it;
+        }
+
+    private:
+        bool _has_incremented = true;
+        Iterator _it;
+        std::string const _root;
+        ConfigTreeNew& _parent;
+    };
+
+
+    /*!
+     * A wrapper around a Boost Iterator for iterating over ranges of values.
+     *
+     * The methods of this class tell the associated (parent) \c ConfigTree object when
+     * a setting has been parsed.
+     */
+    template<typename ValueType>
+    class ValueIterator
+            : public std::iterator<std::input_iterator_tag, ValueType>
+    {
+    public:
+        using Iterator = boost::property_tree::ptree::const_assoc_iterator;
+
+        explicit ValueIterator(Iterator it, std::string const& root,
+                               ConfigTreeNew& parent)
+            : _it(it), _root(root), _parent(parent)
+        {}
+
+        ValueIterator<ValueType>& operator++() {
+            ++_it;
+            _has_incremented = true;
+            return *this;
+        }
+
+        ValueType operator*() {
+            // if this iterator has been incremented since the last dereference,
+            // tell the _parent instance that a setting now has been parsed.
+            if (_has_incremented) {
+                _has_incremented = false;
+                _parent.markVisited<ValueType>(_root);
+            }
+
+            if (_it->second.begin() != _it->second.end()) {
+                _parent.error("Configuration at key " + _root + " has subitems.");
+                return ValueType();
+            }
+
+            auto v = _it->second.get_value_optional<ValueType>();
+
+            if (v) return *v;
+
+            // TODO: change error method
+            _parent.error("Could not get value out of key " + _root + ".");
+            return ValueType();
+        }
+
+        bool operator==(ValueIterator<ValueType> const& other) const {
+            return _it == other._it;
+        }
+
+        bool operator!=(ValueIterator<ValueType> const& other) const {
+            return _it != other._it;
+        }
+
+    private:
+        bool _has_incremented = true;
+        Iterator _it;
+        std::string const _root;
+        ConfigTreeNew& _parent;
+    };
+
+    using PTree = boost::property_tree::ptree;
+
+    //! Type of the function objects used as callbacks.
+    //! The first argument denotes the path in the tree at which an event (warning/error)
+    //! occured, the second argument is the associated message
+    using Callback = std::function<void(const std::string& path,
+                                        const std::string& message)>;
+
+    /*! Creates a new instance wrapping the given Boost Property Tree.
+     *
+     * \param tree the Boost Property Tree to be wrapped
+     * \param onerror callback function to be called on error.
+     * \param onwarning callback function to be called on warning.
```

Comment:
the onerror and onwarning are noew error_cb and warning_cb.
clang complains here:
```
/home/naumov/w/ogs/BaseLib/ConfigTreeNew.h:173:15: warning: parameter 'onerror' not found in the function declaration [-Wdocumentation]
     * \param onerror callback function to be called on error.
              ^~~~~~~
/home/naumov/w/ogs/BaseLib/ConfigTreeNew.h:174:15: warning: parameter 'onwarning' not found in the function declaration [-Wdocumentation]
     * \param onwarning callback function to be called on warning.
              ^~~~~~~~~
```